api/filters/sharing_instance_filter.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filters
+
+import (
+	"encoding/json"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/pkg/web"
+	"github.com/Peripli/service-manager/storage"
+	"github.com/sirupsen/logrus"
+	"net/http"
+)
+
+const SharingInstanceFilterName = "SharingInstanceFilter"
+
+// SharingInstanceFilter validate the un/share request on an existing service instance
+type sharingInstanceFilter struct {
+	storageRepository storage.Repository
+}
+
+func NewSharingInstanceFilter(storageRepository storage.Repository) *sharingInstanceFilter {
+	return &sharingInstanceFilter{
+		storageRepository: storageRepository,
+	}
+}
+
+func (*sharingInstanceFilter) Name() string {
+	return SharingInstanceFilterName
+}
+
+func (sf *sharingInstanceFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	var reqServiceInstance types.ServiceInstance
+	err := util.BytesToObjectNoLabels(req.Body, &reqServiceInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := req.Context()
+	instanceID := req.PathParams["resource_id"]
+
+	// Get instance from database
+	dbPersistedInstanceObject, err := storage.GetObjectByField(ctx, sf.storageRepository, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	persistedInstance := dbPersistedInstanceObject.(*types.ServiceInstance)
+	if persistedInstance.IsShared() && persistedInstance.ServicePlanID != reqServiceInstance.ServicePlanID && reqServiceInstance.Shared == nil {
```

Comment:
why the `&& reqServiceInstance.Shared == nil`?

Reply:
We need to include the condition of  && reqServiceInstance.Shared == nil,
because we don't want to mix scenarios and the order of the validation matters.

The validation you are asking about has to come first and before the validation of:
```
	if reqServiceInstance.Shared == nil {
		return next.Handle(req)
	}
```
This comes right after, so we don't want to skip this handler.

Since the first validation does not handle the flow of un/sharing instances but just making sure the user does not change the plan of a shared instance.
So. it must not include the "shared" property.

The second validation says that we skip the filter if you are not un/sharing the instance.

Reply:
I will try to explain it better inside the code, using comments

Reply:
I renamed the condition to:
```
	if changingPlanOfSharedInstance(persistedInstance, reqServiceInstance) {
		// we don't allow changing plan of shared instance
		return nil, util.HandleInstanceSharingError(util.ErrChangingPlanOfSharedInstance, persistedInstance.ID)
	}
	if reqServiceInstance.Shared == nil {
		return next.Handle(req)
	}
```

Reply:
consider replacing it with `&& len(reqServiceInstance.ServicePlanID) != 0`, this way you don't need to explain anything - if you passed a plan ID and it is different than what we have in DB we fail the request, without making any other assumptions on the payload.

Reply:
Isn't it possible that we pass the same plan-id when updating an instance? 
for example, when renaming the instance, I will also pass:
`
service_plan_id=old_plan_id
`

The request body of the patch service instance contains "service_plan_id"
```
{
  "name": "my-service-instance",
  "service_plan_id": "dsdsdsadsa98898890dasdsa",
  "parameters": {
    "parameter1": "value1",
    "parameter2": "value2"
  },
  "labels": [
    {
      "op": "add",
      "key": "label1",
      "values": [
        "value1",
        "value2"
      ]
    }
  ]
}
```

Reply:
yes, i don't mean to remove the other parts of the if, just replace the last one. so you end up with `if persistedInstance.IsShared() && persistedInstance.ServicePlanID != reqServiceInstance.ServicePlanID && reqServiceInstance.ServicePlanID != ""`